test/e2e/clusterctl_upgrade.go

```diff
@@ -428,7 +430,30 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			Name:                   workloadClusterName,
 		}, input.E2EConfig.GetIntervals(specName, "wait-cluster")...)
 
-		expectedMachineCount := *controlPlaneMachineCount + calculateExpectedWorkerCount(ctx, managementClusterProxy.GetClient(), workloadClusterUnstructured, coreCAPIStorageVersion)
+		var infraMachineKind string
+		var infraMachinePool *unstructured.Unstructured
+		skipMachinePoolMachines := false
+
+		By("Determining if MachinePool machines should be skipped")
+		machinePool := getMachinePool(ctx, managementClusterProxy.GetClient(), workloadClusterNamespace, workloadClusterName)
+		if machinePool != nil {
+			infraMachinePool = getInfraMachinePool(ctx, managementClusterProxy.GetClient(), *machinePool, workloadClusterNamespace, workloadClusterName)
+		}
+
+		if infraMachinePool != nil {
+			// Check if the InfraMachinePool has an infrastructureMachineKind field. If it does not, we should skip checking for MachinePool machines.
+			if err := util.UnstructuredUnmarshalField(infraMachinePool, &infraMachineKind, "status", "infrastructureMachineKind"); err != nil {
+				if errors.Is(err, util.ErrUnstructuredFieldNotFound) {
+					skipMachinePoolMachines = true
+				}
+			}
+		}
+
+		expectedWorkerCount := calculateExpectedWorkerCount(ctx, managementClusterProxy.GetClient(), workloadClusterUnstructured, coreCAPIStorageVersion)
+		expectedMachineCount := *controlPlaneMachineCount
+		if !skipMachinePoolMachines {
+			expectedMachineCount += expectedWorkerCount
```

Comment:
Don't we have to diferentiate between machinepool workers and machinedeployment workers here?

Isn't this something like:

```suggestion
		expectedMachineCount := *controlPlaneMachineCount + expectedMachineDeploymentWorkerCount
		if !skipMachinePoolMachines {
			expectedMachineCount += expectedMachinePoolWorkerCount
```